src/storage/mod.rs

```diff
@@ -980,6 +980,183 @@ impl<E: Engine, L: LockManager, F: KvFormat> Storage<E, L, F> {
+        let quota_limiter = self.quota_limiter.clone();
+        let mut sample = quota_limiter.new_sample(true);
+        with_tls_tracker(|tracker| {
+            tracker.metrics.grpc_process_nanos =
+                stage_begin_ts.saturating_elapsed().as_nanos() as u64;
+        });
+        self.read_pool_spawn_with_busy_check(
+            busy_threshold,
+            async move {
+                let stage_scheduled_ts = Instant::now();
+                let mut key_ranges = vec![];
+                for key in &keys {
+                    key_ranges.push(build_key_range(key.as_encoded(), key.as_encoded(), false));
+                }
+                tls_collect_query_batch(
+                    ctx.get_region_id(),
+                    ctx.get_peer(),
+                    key_ranges,
+                    QueryKind::Get,
+                );
+
+                KV_COMMAND_COUNTER_VEC_STATIC.get(CMD).inc();
+                SCHED_COMMANDS_PRI_COUNTER_VEC_STATIC
+                    .get(priority_tag)
+                    .inc();
+
+                deadline.check()?;
+
+                Self::check_api_version(
+                    api_version,
+                    ctx.api_version,
+                    CMD,
+                    keys.iter().map(Key::as_encoded),
+                )?;
+
+                let command_duration = Instant::now();
+
+                let snap_ctx = prepare_snap_ctx(
+                    &ctx,
+                    &keys,
+                    start_ts,
+                    // buffer_batch_get doesn't read locks from other txns
+                    &TsSet::Empty,
+                    &concurrency_manager,
+                    CMD,
+                )?;
+                let snapshot =
+                    Self::with_tls_engine(|engine| Self::snapshot(engine, snap_ctx)).await?;
+                {
+                    deadline.check()?;
+                    let begin_instant = Instant::now();
+
+                    let stage_snap_recv_ts = begin_instant;
+                    let (result, stats) = Self::with_perf_context(CMD, || {
+                        let _guard = sample.observe_cpu();
+                        let mut reader = MvccReader::new(
+                            snapshot,
+                            Some(ScanMode::Forward),
+                            !ctx.get_not_fill_cache(),
+                        );
+                        // TODO: metrics
+                        // TODO: refactor: reuse functions in PointGetter
+                        let result: Vec<Result<KvPair>> = keys
+                            .into_iter()
+                            .filter_map(|k| {
+                                let pair: Option<std::result::Result<KvPair, _>> =
+                                    match reader.load_lock(&k) {
+                                        Ok(None) => None,
+                                        Ok(Some(lock)) => {
+                                            if lock.ts != start_ts {
```

Comment:
Need to check if it's prewrite lock, perhaps an assertion not is needed for pessimistic lock type.

Reply:
I prefer letting it ignore the pessimistic lock. What do you think?

Reply:
It's better to return error messages and print error logs if the pessimistic locks are to be ignored, so we could find out if the interface is used mistakenly by the client.

Reply:
If key `k` is not touched by a transaction, its `membuffer.get(k)` should return null. If another transaction acquires pessimistic lock on it, the lock is ignored and no errors will be returned.
IMO `BufferBatchGet` is a subset of the generalized `membuffer.get()`, so it should not throw errors that do not exist in the original interface?

Reply:
Yes, here I mean only the returned lock whose `start_ts` equals the `start_ts` of the pipelined transaciton shold not be pessimistic lock type. Locks by concurrent transactions should be ignored.